File: app/src/main/java/com.example.safehome/Publisher/SensorNotification.java
package com.example.safehome.Publisher;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import java.util.ArrayList;
import java.util.List;

/**
 * Concrete implementation of SensorNotification
 * @param <Data>
 */
public class SensorNotification<Data> implements Notification<Data> {
    private static final Logger log = LoggerFactory.getLogger(SensorNotification.class);
    private boolean state;
    private List<Data> listOfData;
    private EventType eventType;


    public SensorNotification(EventType eventType, boolean state, Data data) {
        this.eventType = eventType;
        this.state = state;
        this.listOfData = new ArrayList<>();
        this.listOfData.add(data);
    }

    /**
     * enables notifications
     */
    @Override
    public boolean enable() {
        return false;
    }

    /**
     * disables notifications
     */
    @Override
    public boolean disable() {
        return false;
    }

    /**
     * updates notification status
     *
     * @param data
     */
    @Override
    public boolean update(Data data) {
        return false;
    }

    /**
     * gets state from system
     */
    @Override
    public boolean getState() {
        return false;
    }

    /**
     * sets state in system
     *
     * @param state
     */
    @Override
    public void setState(boolean state) {

    }

    /**
     * sets type in system
     *
     * @param type
     */
    @Override
    public void setType(EventType type) {

    }

    /**
     * gets type from system
     *
     * @return {@linkplain EventType}
     */
    public EventType getType() {
        return this.eventType;
    }

    /**
     * gets data message received from an event
     * @return Data received from an event
     */
    public Data getData() {
        try{
            return this.listOfData.get(listOfData.size()-1);
        }catch (Exception e){
            log.error("{}", e);
        }
        return null;
    }


    /**
     * gets event from system
     */

    public EventType getEvent() {
        return eventType;
    }
}

File: app/src/main/kotlin/com.example.safehome/Publisher/SensorNotification.kt
package com.example.safehome.Publisher


import org.slf4j.LoggerFactory
import java.lang.Exception
import java.util.ArrayList

/**
 * Concrete implementation of SensorNotification
 * @param <Data>
</Data> */
class SensorNotification (
    /**
     * gets event from system
     */
    var event: EventType, private val state: Boolean, data: Data
) : Notification {
    private val listOfData: MutableList<Data>

    init {
        listOfData = ArrayList()
        listOfData.add(data)
    }

    /**
     * enables notifications
     */
    override fun enable(): Boolean {
        return false
    }

    /**
     * disables notifications
     */
    override fun disable(): Boolean {
        return false
    }

    /**
     * updates notification status
     *
     * @param data
     */
    override fun update(data: Data?): Boolean {
        return false
    }

    /**
     * gets state from system
     */
    override fun getState(): Boolean {
        return false
    }

    /**
     * sets state in system
     *
     * @param state
     */
    override fun setState(state: Boolean) {}
    /**
     * gets type from system
     *
     * @return [EventType]
     */
    /**
     * sets type in system
     *
     * @param type
     */
    /**
    override var type: EventType?
        get() = event
        set(type) {this.type = event}
*/
    /**
     * sets type of event
     */
    override fun setType(type: EventType?) {
        if (type != null) {
            this.event = type
        }
    }

    /**
     * gets type of event
     * @return EventType
     */
    override fun getType(): EventType? {
        if (event != null) {
            return this.event
        }
        return null
    }


    /**
     * gets data message received from an event
     * @return Data received from an event
     */
    override fun getData(): Data?{
        try {
            return listOfData[listOfData.size - 1]
        } catch (e: Exception) {
            log.error("{}", e)
        }
        return null
    }

    companion object {
        private val log = LoggerFactory.getLogger(SensorNotification::class.java)
    }
}